File: sdks/python/src/aether_sdk/transaction.py
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from typing import List, Optional

from .types import ensure_hex, ensure_positive_int


@dataclass
class Transaction:
    nonce: int
    sender: str
    sender_public_key: str
    recipient: str
    amount: int
    fee: int
    gas_limit: int
    memo: Optional[str]
    signature: str
    reads: List[str] = field(default_factory=list)
    writes: List[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        ensure_hex(self.sender, field="sender")
        ensure_hex(self.sender_public_key, field="sender_public_key")
        ensure_hex(self.recipient, field="recipient")
        ensure_hex(self.signature, field="signature")
        ensure_positive_int(self.amount, field="amount")
        ensure_positive_int(self.fee, field="fee")
        ensure_positive_int(self.gas_limit, field="gas_limit")
        ensure_positive_int(self.nonce + 1, field="nonce + 1")  # nonce can be 0
        if len(self.signature) < 66:
            raise ValueError("signature must be at least 64 bytes (hex)")
        if not self.writes:
            self.writes.append(self.recipient)

    def hash(self) -> str:
        payload = {
            "nonce": self.nonce,
            "sender": self.sender,
            "sender_public_key": self.sender_public_key,
            "recipient": self.recipient,
            "amount": str(self.amount),
            "fee": str(self.fee),
            "gas_limit": self.gas_limit,
            "memo": self.memo,
            "reads": self.reads,
            "writes": self.writes,
        }
        digest = hashlib.sha256(json.dumps(payload, sort_keys=True).encode())
        return "0x" + digest.hexdigest()

File: sdks/typescript/src/transaction.ts
import { createHash } from "node:crypto";

import type { TransactionFields } from "./types.js";

export class Transaction {
  readonly nonce: number;
  readonly sender: string;
  readonly senderPublicKey: string;
  readonly recipient: string;
  readonly amount: bigint;
  readonly fee: bigint;
  readonly gasLimit: number;
  readonly memo?: string;
  readonly signature: string;
  readonly reads: string[];
  readonly writes: string[];

  constructor(fields: TransactionFields) {
    if (!fields.signature || fields.signature.length < 128) {
      throw new Error("signature must be at least 64 characters");
    }

    this.nonce = fields.nonce;
    this.sender = fields.sender;
    this.senderPublicKey = fields.senderPublicKey;
    this.recipient = fields.recipient;
    this.amount = fields.amount;
    this.fee = fields.fee;
    this.gasLimit = fields.gasLimit;
    this.memo = fields.memo;
    this.signature = fields.signature;
    this.reads = [fields.sender];
    this.writes =
      fields.sender === fields.recipient
        ? [fields.sender]
        : [fields.sender, fields.recipient];
  }

  hash(): string {
    const forHash = {
      nonce: this.nonce,
      sender: this.sender,
      senderPublicKey: this.senderPublicKey,
      recipient: this.recipient,
      amount: this.amount.toString(),
      fee: this.fee.toString(),
      gasLimit: this.gasLimit,
      memo: this.memo,
      reads: this.reads,
      writes: this.writes,
    };
    const hash = createHash("sha256")
      .update(JSON.stringify(forHash))
      .digest("hex");
    return `0x${hash}`;
  }

  toJSON() {
    return {
      nonce: this.nonce,
      sender: this.sender,
      senderPublicKey: this.senderPublicKey,
      recipient: this.recipient,
      amount: this.amount.toString(),
      fee: this.fee.toString(),
      gasLimit: this.gasLimit,
      memo: this.memo,
      signature: this.signature,
      reads: this.reads,
      writes: this.writes,
    };
  }

  toRpcTransaction() {
    return {
      nonce: this.nonce,
      sender: this.sender,
      sender_public_key: this.senderPublicKey,
      recipient: this.recipient,
      amount: this.amount.toString(),
      fee: this.fee.toString(),
      gas_limit: this.gasLimit,
      memo: this.memo,
      reads: this.reads,
      writes: this.writes,
      signature: this.signature,
    };
  }
}
